From f56b90ccc1877be47e87740fe547ae1a76084679 Mon Sep 17 00:00:00 2001
From: Kyle Hsiao <kylehsiao@google.com>
Date: Mon, 22 Jun 2026 20:47:21 -0700
Subject: [PATCH 05/13] Fix out-of-bounds write in NFC activation handling

When an NFCC sends a short RF_INTF_ACTIVATED_NTF packet, a parsing failure can bypass the initialization of a tech parameter union on the stack. This uninitialized stack data is later used as a length parameter for a memcpy, leading to an out-of-bounds write in the NFC process that can corrupt function pointers.

This commit resolves the issue by fully zero-initializing the `evt_data.activate` structure in `nfc_ncif_proc_activate()`, preventing uninitialized stack memory from being passed to the callback path. Furthermore, it adds a bounds check in `nfa_dm_notify_activation_status()` to clamp `nfcid_len` to `NCI_NFCID1_MAX_LEN` when decoding the `NFC_DISCOVERY_TYPE_POLL_A` parameter block, mitigating unbounded writes.

Test: Built libnfc-nci module and ran alint.

Bug: 522373576
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

Change-Id: Iedd07da2dd238a27550c8d8bdfa9ab8d4cd1050f
Merged-In: Iedd07da2dd238a27550c8d8bdfa9ab8d4cd1050f
(cherry picked from commit f974b0fee85ff9910610f913e64df59c827cdaf7)
---
 libnfc-nci/src/nfa/dm/nfa_dm_act.cc | 11 +++++++++++
 libnfc-nci/src/nfc/nfc/nfc_ncif.cc  |  2 +-
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/libnfc-nci/src/nfa/dm/nfa_dm_act.cc b/libnfc-nci/src/nfa/dm/nfa_dm_act.cc
index afea0d2..b2aa543 100644
--- a/libnfc-nci/src/nfa/dm/nfa_dm_act.cc
+++ b/libnfc-nci/src/nfa/dm/nfa_dm_act.cc
@@ -24,6 +24,7 @@
  ******************************************************************************/
 #include <android-base/logging.h>
 #include <android-base/stringprintf.h>
+#include <log/log.h>
 #include <string.h>
 
 #include "nci_hmsgs.h"
@@ -1718,6 +1719,11 @@ void nfa_dm_notify_activation_status(tNFA_STATUS status,
                p_nfcid, nfcid_len);
       } else {
         nfcid_len = p_tech_params->param.pa.nfcid1_len;
+        if (nfcid_len > NCI_NFCID1_MAX_LEN) {
+          android_errorWriteLog(0x534e4554, "522373576");
+          nfcid_len = NCI_NFCID1_MAX_LEN;
+          p_tech_params->param.pa.nfcid1_len = nfcid_len;
+        }
         p_nfcid = p_tech_params->param.pa.nfcid1;
       }
     } else if (p_tech_params->mode == NFC_DISCOVERY_TYPE_POLL_B) {
@@ -1731,6 +1737,11 @@ void nfa_dm_notify_activation_status(tNFA_STATUS status,
       p_nfcid = p_tech_params->param.pi93.uid;
     } else if (p_tech_params->mode == NFC_DISCOVERY_TYPE_POLL_KOVIO) {
       nfcid_len = p_tech_params->param.pk.uid_len;
+      if (nfcid_len > NFC_KOVIO_MAX_LEN) {
+          android_errorWriteLog(0x534e4554, "522373576");
+          nfcid_len = NFC_KOVIO_MAX_LEN;
+          p_tech_params->param.pk.uid_len = nfcid_len;
+      }
       p_nfcid = p_tech_params->param.pk.uid;
     } else {
       nfcid_len = 0;
diff --git a/libnfc-nci/src/nfc/nfc/nfc_ncif.cc b/libnfc-nci/src/nfc/nfc/nfc_ncif.cc
index b60ab41..117e4e0 100644
--- a/libnfc-nci/src/nfc/nfc/nfc_ncif.cc
+++ b/libnfc-nci/src/nfc/nfc/nfc_ncif.cc
@@ -964,7 +964,7 @@ void nfc_ncif_proc_activate(uint8_t* p, uint8_t len) {
 
   nfc_set_state(NFC_STATE_OPEN);
 
-  memset(p_intf, 0, sizeof(tNFC_INTF_PARAMS));
+  memset(&evt_data.activate, 0, sizeof(evt_data.activate));
   // validate packet length should be larger than (rf_disc_id + type +
   // protocol + mode + buff_size + num_buff + length of rf parameters).
   if (plen < 7) {
-- 
2.55.0

